cprover_bindings/src/irep/goto_binary_serde.rs

```diff
@@ -0,0 +1,1339 @@
+        let _is_type = (flags & (1 << 15)) != 0;
+        let _is_property = (flags & (1 << 14)) != 0;
+        let _is_macro = (flags & (1 << 13)) != 0;
+        let _is_exported = (flags & (1 << 12)) != 0;
+        let _is_input = (flags & (1 << 11)) != 0;
+        let _is_output = (flags & (1 << 10)) != 0;
+        let _is_state_var = (flags & (1 << 9)) != 0;
+        let _is_parameter = (flags & (1 << 8)) != 0;
+        let _is_auxiliary = (flags & (1 << 7)) != 0;
+        let _is_binding = (flags & (1 << 6)) != 0; // deprecated
+        let _is_lvalue = (flags & (1 << 5)) != 0;
+        let _is_static_lifetime = (flags & (1 << 4)) != 0;
+        let _is_thread_local = (flags & (1 << 3)) != 0;
+        let _is_file_local = (flags & (1 << 2)) != 0;
+        let _is_extern = (flags & (1 << 1)) != 0;
+        let _is_volatile = (flags & 1) != 0;
+        let _is_volatile = (flags & 0x1) != 0;
+
+        let shifted_flags = flags >> 16;
+
+        if shifted_flags != 0 {
+            return Err(Error::new(
+                ErrorKind::Other,
+                "incorrect binary format: set bits remain in decoded symbol flags",
+            ));
+        }
+        Ok(())
+    }
+
+    /// Reads a whole SymbolTable from the byte stream.
+    fn read_symbol_table(&mut self) -> io::Result<()> {
+        // Write symbol table size
+        let symbol_table_len = self.read_usize_varenc()?;
+
+        // Write symbols
+        for _ in 0..symbol_table_len {
+            self.read_symbol()?;
+        }
+
+        Ok(())
+    }
+
+    /// Reads an empty function map from the byte stream.
+    fn read_function_map(&mut self) -> io::Result<()> {
+        let goto_function_len = self.read_usize_varenc()?;
+        Self::expect(goto_function_len, 0)?;
+        Ok(())
+    }
+
+    /// Reads a GOTO binary header from the byte stream.
+    fn read_header(&mut self) -> io::Result<()> {
+        // Read header
+        Self::expect(self.read_u8().unwrap(), 0x7f)?;
+        Self::expect(self.read_u8().unwrap(), b'G')?;
+        Self::expect(self.read_u8().unwrap(), b'B')?;
+        Self::expect(self.read_u8().unwrap(), b'F')?;
+
+        // Read goto binary version
+        let goto_binary_version = self.read_usize_varenc()?;
+        if goto_binary_version != 5 {
+            return Err(Error::new(
+                ErrorKind::Other,
+                format!(
+                    "unsupported GOTO binary version: {}. Supported version: {}",
+                    goto_binary_version, 5
+                ),
+            ));
+        }
+        Ok(())
+    }
+
+    /// Read a GOTO binary file from the byte stream.
+    fn read_file(&mut self) -> io::Result<()> {
+        self.read_header()?;
+        self.read_symbol_table()?;
+        self.read_function_map()?;
+        Ok(())
+    }
+}
+
+////////////////////////////////////////
+//// Dynamic memory usage computation
+////////////////////////////////////////
+
+impl DynamicUsage for NumberedIrep {
```

Comment:
Does this need to be in production code, or should it be guarded by development mode (e.g. using https://doc.rust-lang.org/reference/conditional-compilation.html#debug_assertions)?

Reply:
which mode should I use ? `#[cfg(debug_assertions)]` or `#[cfg(test)]` ?